### ISSUE TITLE
Automatically delete plugin cache at the slightest sign of trouble (BUG: Uninstalled plugins are not uncached)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10625]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| justinj reported|
|-|-|
|Trac ID|trac#926|
|Type|defect|
|Created|2005-03-17 00:14:58Z|

Attachments:

* [twisted-plugin.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/89d/89da3d8a3c24dbd3e5e96f7f3896866f5e5eb878/e5dccbbd93479ea2f966c97ffd24557e47165db1.patch) (1827 bytes) - added by marienz on 2006-06-07 12:54:44Z - the same patch but with a test

<details><summary>Searchable metadata</summary>

```
trac-id__926 926
type__defect defect
reporter__justinj justinj
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1111018498000000 1111018498000000
changetime__1149981060000000 1149981060000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__itamarst cc__justinj cc__marienz
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
I got the following error from the TwistedSumo tar ball and from SVN as of 
March 16, 12:00 CST.  I'm running on Fedora Core 2 with Python 2.4.  Completely 
removed old Twisted installations before installing.

$ mktap names --pyzone justinj.org
which: no dia in 
(/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/justinj/bin
)
Traceback (most recent call last):
  File "/usr/local/bin/mktap", line 19, in ?
    run()
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", line 
186, in run
    options.parseOptions(sys.argv[1:])
  File "/usr/local/lib/python2.4/site-packages/twisted/python/usage.py", line 
197, in parseOptions
    self.postOptions()
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", line 
168, in postOptions
    self.tapLookup = loadPlugins(debug, progress)
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", line 
75, in loadPlugins
    for plug in plugins:
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 202, in 
getPlugIns
    adapted = interface(plugin, None)
  File "/usr/local/lib/python2.4/site-packages/zope/interface/interface.py", 
line 658, in __call__
    adapter = conform(self)
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 77, in 
__conform__
    return self.load()
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 72, in 
load
    return namedAny(self.dropin.moduleName + '.' + self.name)
  File "/usr/local/lib/python2.4/site-packages/twisted/python/reflect.py", line 
389, in namedAny
    obj = getattr(obj, n)
AttributeError: 'module' object has no attribute 'twisted_news'
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
what's the deally, yo?
I remember talking to exarkun about this, and I think I remember some 
indication that he was fixing / had fixed this. comment, exarkun?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
This is an old problem, so I think we could ignore it until after the release.

However, I have a patch that causes lore to not use plugins for the things it
supplies, which gets rid of this undesirable behavior.

I can apply it now or after 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Can you attach the patch? I'm not hugely concerned but it'd be cool to have 
a fix before 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Checked into a branch at /branches/exarkun/lorefix/

Feel free to merge if you feel it is appropriate.
Might be good to talk to someone who actually uses these features of lore to see
if they broke anything.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
This bug is preventing me from running mktap names at all.  Shouldn't something 
like that be applied for this release?  I never tried to use names before, so I 
can't attest as to whether this worked before.  But the fact that it doesn't 
work at all if dia isn't installed seems like a big deal to me.  Am I missing 
something?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
justin, the "which dia" error had nothing to do with the actual problem you
had with mktap names -- which dia was just spurious useless output, while 
the problem you had with names was that the names plugin file wasn't 
included in the pre-release you were trying to install. That latter problem 
has been fixed since a4.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Additionally, there is no `mktap names' command.  There is only `mktap dns'.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Doh!  I don't know how I missed that!  So with the latest svn, 'mktap dns' 
works for me, with the following output.

$ mktap dns --pyzone justinj.org
Traceback (most recent call last):
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", 
line 193, in run
    options.parseOptions(sys.argv[1:])
  File "/usr/local/lib/python2.4/site-packages/twisted/python/usage.py", 
line 197, in parseOptions
    self.postOptions()
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", 
line 175, in postOptions
    self.tapLookup = loadPlugins(debug, progress)
  File "/usr/local/lib/python2.4/site-packages/twisted/scripts/mktap.py", 
line 75, in loadPlugins
    for plug in plugins:
--- &lt;exception caught here> ---
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 
203, in getPlugIns
    adapted = interface(plugin, None)
  File "/usr/local/lib/python2.4/site-packages/zope/interface/
interface.py", line 658, in __call__
    adapter = conform(self)
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 77, 
in __conform__
    return self.load()
  File "/usr/local/lib/python2.4/site-packages/twisted/plugin.py", line 72, 
in load
    return namedAny(self.dropin.moduleName + '.' + self.name)
  File "/usr/local/lib/python2.4/site-packages/twisted/python/reflect.py", 
line 389, in namedAny
    obj = getattr(obj, n)
exceptions.AttributeError: 'module' object has no attribute 'twisted_news'
$ ls
dns.tap  justinj.org
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Right now the logging is intentionally extremely verbose so we notice any
problems.  This traceback is due to the fact that a cache file is in use which
was generated from an installed plugin which no longer exists.  I guess we need
to notice errors from removed plugins and invalidate the cache.  We could do
this either by noticing an AttributeError from namedAny or by examining the
mtime of the containing directory.

Another possibilities is to requiring the sysadmin to remove the cache when
uninstalling packages, and perhaps to shorten the reporting of this particular
error case (to something like "Missing plugin: /full/path/to/twisted_news.py").
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
We can't really check for this, since the whole point of that cache is to avoid
a zillon stat()s.  However, we can at least formalize the process of deleting
the dropin.cache on package uninstallation.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|marienz commented|
|-|-|

```
#!html
<pre>
I suspect I'm missing something, but it looks to me like plugin.py is already
calling stat on every python-related file in the dir the dropin.cache is in when
you call getPlugins. The attached patch doesn't seem to break any tests and does
seem to solve the problem, although it's not heavily tested.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

This patch is good.  I want to apply it.

It's waiting on a test.

Eventually I'll write a test but anyone who wants to can add one and accellerate this getting merged :)

**Comment by twisted-trac** [comment]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|marienz commented|
|-|-|

Something like this? It fails with just the test bit applied and passes with both bits applied, so it's probably at least in the right direction.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Yes, exactly like that!  Thanks.  I will apply this now ;).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

Oh, and in the vein that there should be at least some comments on reviews: it would be nice if that test wouldn't write to the filesystem, and would instead manipulate transient Python state.

That is not enough to block the merge though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17162]) Remove modules found not to be installed from the plugin cache.

This was a long-standing bug in twisted.plugin.  Now you can uninstall stuff
without breaking getPlugins lookups!

Author: marienz

Reviewer: glyph

Fixes [#10625](https://github.com/twisted/twisted/issues/10625)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
